include/ofi_recvwin.h

```diff
@@ -0,0 +1,152 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * This utility provides a sliding receive window implementation. It uses a
+ * circular queue to order incoming messages based on the message ID in the
+ * transport protocol.
+ */
+
+#if !defined(FI_RECVWIN_H)
+#define FI_RECVWIN_H
+
+#include "config.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <ofi.h>
+#include <ofi_rbuf.h>
+
+#define OFI_DECL_RECVWIN_BUF(entrytype, name)				\
+OFI_DECLARE_CIRQUE(entrytype, recvwin_cirq);				\
+struct name {								\
+	uint64_t exp_msg_id;						\
+	unsigned int win_size;						\
+	struct recvwin_cirq *pending;					\
+};									\
+DECLARE_FREESTACK(struct name, name ##_fs);				\
+									\
+static inline struct name *						\
+ofi_recvwin_buf_alloc(struct name ##_fs *fs, unsigned int size)		\
+{									\
+	struct name *recvq;						\
+	assert(size == roundup_power_of_two(size));			\
+	if (freestack_isempty(fs))					\
+		return NULL;						\
+	recvq		= freestack_pop(fs);				\
+	recvq->exp_msg_id = 0;						\
+	recvq->win_size	= size;						\
+	recvq->pending	= recvwin_cirq_create(recvq->win_size);		\
+	return recvq;							\
+}									\
+									\
+static inline void							\
+ofi_recvwin_free(struct name ##_fs *fs, struct name *recvq)		\
+{									\
+	recvwin_cirq_free(recvq->pending);				\
+	freestack_push(fs, recvq);					\
+}									\
+									\
+static inline void							\
+ofi_recvwin_exp_inc(struct name *recvq)					\
+{									\
+	recvq->exp_msg_id++;						\
+}									\
+									\
+static inline bool							\
+ofi_recvwin_is_exp(struct name *recvq, uint64_t id)			\
+{									\
+	return recvq->exp_msg_id == id;					\
+}									\
+									\
+static inline bool							\
+ofi_recvwin_is_delayed(struct name *recvq, uint64_t id)			\
+{									\
+	return recvq->exp_msg_id > id;					\
+}									\
+									\
+static inline bool							\
+ofi_recvwin_is_allowed(struct name *recvq, uint64_t id)			\
+{									\
+	return id >= recvq->exp_msg_id && id < recvq->win_size;		\
+}									\
+									\
+static inline bool							\
+ofi_recvwin_is_empty(struct name *recvq)				\
+{									\
+	return ofi_cirque_isempty(recvq->pending);			\
+}									\
+									\
+static inline int							\
+ofi_recvwin_queue_msg(struct name *recvq, entrytype *msg,		\
+		   uint64_t expected_id, uint64_t id)			\
```

Comment:
I don't understand the need for passing in expected_id and why it would differ from recvq->exp_msg_id.

Reply:
With the original patch, the receiver could end up expecting a newer msg_id than what the sliding window was tracking and the expected_id in the insert was to implicitly update internal state based on what it was expecting at that point in time. Now they could just call `ofi_recvwin_exp_inc()` anytime they do something outside this API. Will fix.